File: packages/sdk/src/http.ts
import { SandkilnApiError } from "./errors.js";
import type { ApiErrorBody } from "./types.js";

export interface RequestOptions {
  baseUrl: string;
  method: "GET" | "POST" | "DELETE";
  path: string;
  body?: unknown;
  authToken?: string;
}

export async function request<T>(options: RequestOptions): Promise<T> {
  const headers: Record<string, string> = {};
  if (options.body !== undefined) {
    headers["content-type"] = "application/json";
  }
  if (options.authToken !== undefined) {
    headers.authorization = `Bearer ${options.authToken}`;
  }

  const response = await fetch(`${options.baseUrl}${options.path}`, {
    method: options.method,
    headers: Object.keys(headers).length > 0 ? headers : undefined,
    body: options.body === undefined ? undefined : JSON.stringify(options.body),
  });

  if (!response.ok) {
    throw new SandkilnApiError(response.status, await extractErrorMessage(response));
  }

  if (response.status === 204) {
    return undefined as T;
  }

  return (await response.json()) as T;
}

async function extractErrorMessage(response: Response): Promise<string> {
  try {
    const body = (await response.json()) as ApiErrorBody;
    if (typeof body.error === "string") {
      return body.error;
    }
  } catch {
    // Non-JSON error body (e.g. a proxy in front of the daemon); fall back below.
  }
  return response.statusText || `request failed with status ${response.status}`;
}

File: packages/python/src/sandkiln/_http.py
import json
import urllib.error
import urllib.request
from typing import Any

from .errors import SandkilnApiError


def request(
    base_url: str,
    method: str,
    path: str,
    auth_token: str | None = None,
    body: Any = None,
) -> Any:
    headers: dict[str, str] = {}
    data: bytes | None = None
    if body is not None:
        data = json.dumps(body).encode("utf-8")
        headers["Content-Type"] = "application/json"
    if auth_token is not None:
        headers["Authorization"] = f"Bearer {auth_token}"

    req = urllib.request.Request(f"{base_url}{path}", data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req) as response:
            return _parse_body(response)
    except urllib.error.HTTPError as error:
        raise SandkilnApiError(error.code, _extract_error_message(error)) from None
    except urllib.error.URLError as error:
        raise SandkilnApiError(0, f"could not reach {base_url}: {error.reason}") from None


def _parse_body(response: Any) -> Any:
    raw = response.read()
    if not raw:
        return None
    return json.loads(raw)


def _extract_error_message(error: urllib.error.HTTPError) -> str:
    try:
        body = json.loads(error.read())
        if isinstance(body, dict) and isinstance(body.get("error"), str):
            return body["error"]
    except (json.JSONDecodeError, UnicodeDecodeError):
        pass
    return error.reason or f"request failed with status {error.code}"
